src/Formatter/DateTimeFormatContext.php

```diff
@@ -0,0 +1,142 @@
+        $self->addField(Field\Year::NAME, (string) $localDateTime->getDate()->getYear());
+        $self->addField(Field\HourOfDay::NAME, (string) $localDateTime->getTime()->getHour());
+        $self->addField(Field\MinuteOfHour::NAME, (string) $localDateTime->getTime()->getMinute());
+        $self->addField(Field\SecondOfMinute::NAME, (string) $localDateTime->getTime()->getSecond());
+        $self->addField(Field\NanoOfSecond::NAME, (string) $localDateTime->getTime()->getNano());
+        $self->addField(Field\FractionOfSecond::NAME, (string) $localDateTime->getTime()->getNano());
+
+        return $self;
+    }
+
+    public static function ofZonedDateTime(ZonedDateTime $zonedDateTime): self
+    {
+        $self = new self($zonedDateTime);
+        $self->addField(Field\DayOfMonth::NAME, (string) $zonedDateTime->getDate()->getDay());
+        $self->addField(Field\DayOfWeek::NAME, (string) $zonedDateTime->getDate()->getDayOfWeek()->getValue());
+        $self->addField(Field\DayOfYear::NAME, (string) $zonedDateTime->getDate()->getDayOfYear());
+        $self->addField(Field\WeekOfYear::NAME, (string) $zonedDateTime->getDate()->getYearWeek()->getWeek());
+        $self->addField(Field\MonthOfYear::NAME, (string) $zonedDateTime->getDate()->getMonth());
+        $self->addField(Field\Year::NAME, (string) $zonedDateTime->getDate()->getYear());
+        $self->addField(Field\HourOfDay::NAME, (string) $zonedDateTime->getTime()->getHour());
+        $self->addField(Field\MinuteOfHour::NAME, (string) $zonedDateTime->getTime()->getMinute());
+        $self->addField(Field\SecondOfMinute::NAME, (string) $zonedDateTime->getTime()->getSecond());
+        $self->addField(Field\NanoOfSecond::NAME, (string) $zonedDateTime->getTime()->getNano());
+        $self->addField(Field\FractionOfSecond::NAME, (string) $zonedDateTime->getTime()->getNano());
+        $self->addField(Field\TimeZoneOffsetHour::NAME, sprintf('%d', floor(abs($zonedDateTime->getTimeZoneOffset()->getTotalSeconds()) / LocalTime::SECONDS_PER_HOUR)));
+        $self->addField(Field\TimeZoneOffsetMinute::NAME, (string) ((abs($zonedDateTime->getTimeZoneOffset()->getTotalSeconds()) % LocalTime::SECONDS_PER_HOUR) / LocalTime::SECONDS_PER_MINUTE));
+        $self->addField(Field\TimeZoneOffsetSign::NAME, $zonedDateTime->getTimeZoneOffset()->getTotalSeconds() === 0 ? 'Z' : ($zonedDateTime->getTimeZoneOffset()->getTotalSeconds() > 0 ? '+' : '-'));
+        $self->addField(Field\TimeZoneOffsetTotalSeconds::NAME, (string) $zonedDateTime->getTimeZoneOffset()->getTotalSeconds());
+        $self->addField(Field\TimeZoneRegion::NAME, $zonedDateTime->getTimeZone()->getId());
+
+        return $self;
+    }
+
+    public function addField(string $name, string $value): void
+    {
+        $this->fields[$name][] = $value;
+    }
+
+    public function hasField(string $name): bool
+    {
+        return isset($this->fields[$name]) && $this->fields[$name];
```

Comment:
what does `&& $this->fields[$name];` check for? a non-empty-string?
maybe better to be explicit for readability? `&& '' !== $this->fields[$name];`